nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,159 @@
+        this.incomingMessages = incomingMessages;
+        this.logger = logger;
+
+    }
+
+    @Override
+    public MessageHandler create(MessageContext messageContext) {
+        return new Handler(messageContext, incomingMessages, logger);
+    }
+
+    class Handler implements MessageHandler {
+        final MessageContext messageContext;
+        String from;
+        String recipient;
+        ByteArrayOutputStream messageData;
+
+        private CountDownLatch latch;
+
+        public Handler(MessageContext messageContext, LinkedBlockingQueue<SmtpEvent> incomingMessages, ComponentLog logger){
+            this.messageContext = messageContext;
+            this.latch =  new CountDownLatch(1);
+        }
+
+        @Override
+        public void from(String from) throws RejectException {
+            // TODO: possibly whitelist senders?
+            this.from = from;
+        }
+
+        @Override
+        public void recipient(String recipient) throws RejectException {
+            // TODO: possibly whitelist receivers?
+            this.recipient = recipient;
+        }
+
+        @Override
+        public void data(InputStream inputStream) throws RejectException, TooMuchDataException {
+            // Start counting the timer...
+
+            StopWatch watch = new StopWatch(true);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            final long serverTimeout = TimeUnit.MILLISECONDS.convert(messageContext.getSMTPServer().getConnectionTimeout(), TimeUnit.MILLISECONDS);
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+
+            int rd;
+
+            try {
+                while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1 ) {
+                    baos.write(buffer, 0, rd);
+                    if (baos.getBufferLength() > server.getMaxMessageSize() ) {
+                        throw new TooMuchDataException("Data exceeds the amount allowed.");
+                    }
+                }
+                baos.flush();
+            } catch (IOException e) {
+                throw new DropConnectionException(450, "Unexpected error processing your message. ");
+            }
+
+            this.messageData = baos;
+
+            X509Certificate[] certificates = new X509Certificate[]{};
+
+            String remoteIP = messageContext.getRemoteAddress().toString();
+            String helo = messageContext.getHelo();
+            if (messageContext.getTlsPeerCertificates() != null ){
+                certificates = (X509Certificate[]) messageContext.getTlsPeerCertificates().clone();
+            }
+
+            SmtpEvent message = new SmtpEvent(remoteIP, helo, from, recipient, certificates, messageData, latch);
+            // Try to queue the message back to the NiFi session
+            try {
+                incomingMessages.offer(message, serverTimeout, TimeUnit.MILLISECONDS);
```

Comment:
This should take into account the currently elapsed time (like the latch does below)